src/rust/lib/logger/src/lib.rs

```diff
@@ -35,112 +39,61 @@ impl<F: Fn() -> S, S:Str> LogMsg for F {
 }
 
 
+
 // ==============
 // === Logger ===
 // ==============
 
-#[derive(Clone,CloneRef,Debug,Default)]
-pub struct Logger {
-    pub path: Rc<String>,
-}
+/// Interface common to all loggers.
+pub trait AnyLogger {
+    /// Path that is used as an unique identifier of this logger.
+    fn path(&self) -> &str;
 
-#[allow(dead_code)]
-impl Logger {
-    pub fn new<T:Str>(path:T) -> Self {
-        let path = Rc::new(path.into());
-        Self {path}
-    }
+    /// Creates a new logger. Path should be a unique identifier for this logger.
+    fn new(path:impl Str) -> Self;
 
-    pub fn sub<T:Str>(&self, path: T) -> Self {
-        if self.path.is_empty() {
-            Self::new(path)
-        } else {
-            Self::new(format!("{}.{}", self.path, path.as_ref()))
-        }
+    /// Creates a new logger with this logger as a parent.
+    fn sub(logger:&impl AnyLogger, path:impl Str) -> Self where Self:Sized {
+        Self::new(format!("{}.{}", logger.path(), path.as_ref()))
```

Comment:
This loses the piece of logic that if previous logger was empty (e.g. was `default`) its sublogger behaves as if root, rather than having a leading dot in path